File: leetcode/medium/08-StringToInteger.py

class Solution:
    def myAtoi(self, s: str) -> int:
        MIN = -2147483648
        MAX = 2147483647
        
        res = 0
        whiteSpace = True
        signNumber = True
        isNegative = False
        for i in range(0, len(s)):
            if s[i] == ' ' and whiteSpace:
                continue
            elif (s[i] == '+' or s[i] == '-') and signNumber:
                whiteSpace = False
                signNumber = False
                if s[i] == '-':
                    isNegative = True
            elif (s[i] >= '0' and s[i] <= '9'):
                whiteSpace = False
                signNumber = False
                res = (res*10)+int(s[i])
                if (res > MAX):
                    break
            else:
                break
        
        if (isNegative):
            res *= -1
        
        if (res <= MIN):
            return MIN
        elif (res >= MAX):
            return MAX
        return res


solve = Solution()

s = "        -123 "
res = solve.myAtoi(s)
print(res)

File: leetcode/medium/08-StringToInteger.cpp
#include <bits/stdc++.h>
using namespace std;

#define lli     long long int


int myAtoi(string s) {
    lli res = 0;
    bool whiteSpace = true,
        signNumber = true,
        isNegative = false;

    for (int i=0; i<s.size(); i++) {
        if (s[i] == ' ' && whiteSpace) { continue; }
        else if ((s[i] == '+' || s[i] == '-') && signNumber) {
            if (s[i] == '-') isNegative = true;

            whiteSpace = false; signNumber = false;
        }
        else if (s[i] >= '0' && s[i] <= '9') {
            res = (res*10) + (s[i] - '0');
            if (res > INT_MAX) break;

            whiteSpace = false; signNumber = false;
        }
        else break;
    }

    if (isNegative) res *= -1;

    if (res >= INT_MAX) return INT_MAX;
    else if (res <= INT_MIN) return INT_MIN;

    return res;
}

int main() {
    string s = "         -123";

    lli res = myAtoi(s);
    cout << res << endl;
    
    return 0;
}
